src/discordcr/cache.cr

```diff
@@ -1,272 +1,66 @@
-    def resolve_dm_channel(recipient_id : UInt64 | Snowflake) : UInt64
-      recipient_id = recipient_id.to_u64
-      @dm_channels.fetch(recipient_id) do
-        channel = @client.create_dm(recipient_id)
-        cache(Channel.new(channel))
-        channel.id
-      end
-    end
-
-    # Resolves the current user's profile. Requires no parameters since the
-    # endpoint has none either. If there is a gateway connection this should
-    # always be cached.
-    def resolve_current_user : User
-      @current_user ||= @client.get_current_user
-    end
-
-    # Deletes a user from the cache given its *ID*.
-    def delete_user(id : UInt64 | Snowflake)
-      @users.delete(id.to_u64)
-    end
-
-    # Deletes a channel from the cache given its *ID*.
-    def delete_channel(id : UInt64 | Snowflake)
-      @channels.delete(id.to_u64)
-    end
-
-    # Deletes a guild from the cache given its *ID*.
-    def delete_guild(id : UInt64 | Snowflake)
-      @guilds.delete(id.to_u64)
-    end
-
-    # Deletes a member from the cache given its *user_id* and the *guild_id* it
-    # is on.
-    def delete_member(guild_id : UInt64 | Snowflake, user_id : UInt64 | Snowflake)
-      guild_id = guild_id.to_u64
-      user_id = user_id.to_u64
-      @members[guild_id]?.try &.delete(user_id)
+  abstract class Cache(K, V)
+    class Error < Exception
     end
 
-    # Deletes a role from the cache given its *ID*.
-    def delete_role(id : UInt64 | Snowflake)
-      @roles.delete(id.to_u64)
-    end
+    abstract def resolve?(key : K)
 
-    # Deletes a DM channel with a particular user given the *recipient_id*.
-    def delete_dm_channel(recipient_id : UInt64 | Snowflake)
-      @dm_channels.delete(recipient_id.to_u64)
+    def fetch(key : K) : V
+      resolve?(key) || cache(key, yield)
     end
 
-    # Deletes the current user from the cache, if that will ever be necessary.
-    def delete_current_user
-      @current_user = nil
+    def resolve(key : K) : V
+      resolve?(key) || raise Error.new("Cache member not found: #{key}")
     end
 
-    # Adds a specific *user* to the cache.
-    def cache(user : User)
-      @users[user.id.to_u64] = user
-    end
+    abstract def cache(key : K, value : V)
 
-    # Adds a specific *channel* to the cache.
-    def cache(channel : Channel)
-      @channels[channel.id.to_u64] = channel
-    end
+    abstract def remove(key : K)
 
-    # Adds a specific *guild* to the cache.
-    def cache(guild : Guild)
-      @guilds[guild.id.to_u64] = guild
-    end
+    abstract def each(&block : Tuple(K, V) ->)
+  end
 
-    # Adds a specific *member* to the cache, given the *guild_id* it is on.
-    def cache(member : GuildMember, guild_id : UInt64 | Snowflake)
-      guild_id = guild_id.to_u64
-      local_members = @members[guild_id] ||= Hash(UInt64, GuildMember).new
-      local_members[member.user.id.to_u64] = member
-    end
+  class MemoryCache(K, V) < Cache(K, V)
+    include Enumerable(Tuple(K, V))
```

Comment:
Why?

Reply:
Ah yeah, I had wanted `cache.find` available etc. but of course you can just do `cache.each.find`. :+1:

Reply:
If you want this, then put `include Enumerable({K, V})` in the abstract `Cache`, not each individual instance. Otherwise you can't use it.

Reply:
Okay, so I've done that. (`include Enumberable({K, V})` in `Cache`)

In your other review comment, if you mean to write:

```cr
abstract def each(&block : K, V ->)
```

And use just `: K, V ->` everywhere else, this works fine for `each`, however, there's some behaviour with `.find` I don't understand.

```cr
client.guild_cache.each do |id, guild|
  # Works, OK
end

# Doesn't compile: too many block arguments (given 2, expected maximum 1)
client.guild_cache.find do |id, guild|
  # ..
end

# Only the ID is yielded:
value = client.guild_cache.find do |id|
  # Works, OK
end

p typeof(value) # => Snowflake | Nil
```

This problem doesn't occur and all examples work if I use `&block : Tuple(K, V) ->` everywhere.

```cr
value = client.guild_cache.find do |id, guild|
  # OK
end

p typeof(value) # => (Tuple(Snowflake, Guild) | Nil)
```

Reply:
Yeah, if you implement `Enumerable` you need to use `{K, V} ->`.